### ISSUE TITLE
Threads and synchronization facilities

### ISSUE THREAD

**Issue by junkblocker** [Hammerspoon/hammerspoon#911]
The level of concurrent execution supported with Timers (and Tasks) in HS is unclear to me. Are they real threads or does only one thread run at any given time or are they time sliced etc.? One of the things I want to write with HS is a download manager with a queue where atomic additions and removal from the queue are needed with some sort of mutex controls. Is that solvable with current facilities? If not, I'd like to suggest/request these features. Thanks!

**Comment by asmagill** [comment]
tl;dr -- Hammerspoon is effectively single threaded and function callbacks are queued as events which the application event loop executes as it gets to them.

While some of the modules in Hammerspoon use alternate threads for some of their operations that do not require direct interaction with Hammerspoon or Lua/LuaSkin, Hammerspoon should be considered a single threaded environment... all Lua execution occurs on the primary thread and things like timer and other function callbacks are queued events which the Hammerspoon run-loop executes in a (for the most part) first-in-first-out manner.  It's why we try not to use blocking activity unless absolutely necessary (for example, it's why `hs.wifi.backgroundScan` is preferred over `hs.wifi.availableNetworks`)

You can use co-routines within Lua code if your code can be written in a chunked manner that supports yielding at regular intervals so that a scheduler can mimic concurrent-like processing, but this can't be done with any Lua function which acts as a callback for a Hammerspoon module -- they don't support yield/resume.  I've occasionally considered seeing how difficult this would be to change, but haven't really done much along these lines yet.

I've also experimented with a module which allows you to run Lua in separate threads and has support for a few of the Hammerspoon modules using a wrapped constructor for LuaSkin... it's probably buggy, has very incomplete Hammerspoon specific support, is at least one or two releases behind Hammerspoon's core (for the modules which have been translated), and not recommended for the feint of heart, but you're welcome to take a look at it and/or contribute to it, if you like... check out https://github.com/asmagill/hs._asm.luathread/ if curious, but I don't think in its present form that it will do what you want.

**Comment by junkblocker** [comment]
Thanks, @asmagill . So, is it like co-routines currently where only one is running at a given time? Could you describe how tasks fit into that too since multiple tasks seem to run and make progress continuously? This details would be great to have on hammerspoon.org too (or maybe somehow fitted in the current docs).

**Comment by heptal** [comment]
@junkblocker The `hs.task`s run in background threads, but invoke their callbacks/update the Lua environment on the main thread using `dispatch_sync(dispatch_get_main_queue(), callback_block)`

**Comment by junkblocker** [comment]
Thanks, @heptal !

**Comment by cmsj** [comment]
@junkblocker all Lua work has to happen on a single thread because the Lua interpreter is not multi-threaded. Hammerspoon modules that are using OS X APIs may well be doing all their work on another thread (as hs.task does), but any time they need to talk to Lua, they pause while that happens on the main thread.

For your download manager, you could be using hs.http's async gets, which happen on separate threads, with all the control work happening in Lua (thus on the main thread). So, you could be running multiple async HTTP operations, with their callbacks/polling happening on the main thread. Not the highest performance solution possible, but it ought to be pretty reasonable.

**Comment by cmsj** [comment]
(I'm going to close this out because I don't think there is any work to be done in Hammerspoon. Please re-open if you disagree!)
